src/Entities/Entity.cpp

```diff
@@ -1032,40 +1032,18 @@ void cEntity::HandlePhysics(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 	}
 
 	// Get water direction
-	Direction WaterDir = m_World->GetWaterSimulator()->GetFlowingDirection(BlockX, BlockY, BlockZ);
+	Vector3f WaterDir = m_World->GetWaterSimulator()->GetFlowingDirection(BlockX, BlockY, BlockZ);
 
 	m_WaterSpeed *= 0.9;  // Reduce speed each tick
 
-	switch (WaterDir)
+	if (WaterDir.x != 0.0f)
```

Comment:
No need for this condition - if the `x` is zero, then adding it will have no effect anyway.
Just `Normalize` the `WaterDir` and then add it to `m_WaterSpeed` directly.

Reply:
I think the condition is needed.  This way the water speed will carry momentum through sections of still water but if the condition weren't there it would be immediately set to 0.

Reply:
Not really. I propose changing the whole block of lines 1037 - 1047 (of the new code) with a single line:
```cpp
m_WaterSpeed = m_WaterSpeed * 0.9 + WaterDir.NormalizeCopy() * 0.4;
```
This is equivalent to the whole code block. Basically if `WaterDir.x == 0`, then adding it to `m_WaterSpeed` will have no effect anyway, so the `if` can go away, and the whole calculation can be `Vector3`-ized.

Reply:
At least use a proper float-comparison - you cannot compare a `float` value for equality, because you'll almost never reach it due to rounding errors. Use `if (std::abs(WaterDir.x) < 0.0005)`